add ceedling 1.0.0 compatibility, add multi project setup capability

simeon-s1 opened the pull request.

simeon-s1 committed:
add ceedling 1.0.0 compatibility, add multi project setup capability

simeon-s1 committed:
add deepmerge yml files

simeon-s1 committed:
update README.md

gsmithufl commented:
I've tested this and I'm able to step debug from the GUI menu and run tests. However, I'm having an issue where tests are showing as failed in the GUI menu. In Test Explorer the output shows as on `stderr`:

```
Ceedling operations completed in 391 milliseconds

stderr:

-----------------------
✅ OVERALL TEST SUMMARY
-----------------------
TESTED:  2
PASSED:  2
FAILED:  0
IGNORED: 0
```

I currently have the project.yaml plugins setup as follows:

```yml
:plugins:
  :load_paths: []
  :enabled:
    - module_generator               # handy for quickly creating source, header, and test templates
    - gcov                           # test coverage using gcov. Requires gcc, gcov, and a coverage analyzer like gcovr

    # Report options (You'll want to choose one stdout option, but may choose multiple stored options if desired)
    #- report_build_warnings_log
    # - report_tests_gtestlike_stdout
    #- report_tests_ide_stdout
    - report_tests_log_factory
    - report_tests_pretty_stdout
    #- report_tests_raw_output_log
    # - report_tests_teamcity_stdout

# Specify which reports you'd like from the log factory
:report_tests_log_factory:
  :reports:
    # - json
    - junit
    # - cppunit
    # - html
```

I think the `README.md` probably needs one more combing through. I believe the section on `ceedlingExplorer.testCommandArgs` probably needs to be updated because I think they got rid of the `utils:gcov` command extension.

Also, the `packages.json` section `"publisher": "futavis",` probably isn't correct for this repo.

Overall, thanks a lot for making these updates! Hopefully this repo owner will show up.

jerdwys commented:
@gsmithufl  I believe you need to enable cppunit under report_tests_log_factory.

From the [breaking changes](https://github.com/ThrowTheSwitch/Ceedling/blob/test/ceedling_0_32_rc/docs/BreakingChanges.md#consolidation-of-test-report-generation-plugins-%EF%B8%8F-report_tests_log_factory):
In addition, all references and naming connected to the previous xml_tests_report plugin have been updated to refer to CppUnit rather than generic XML as this is the actual format of the report that is processed.

jerdwys committed:
The filename changed for the xml report with Ceedling 1.0.0.  Tested this with Ceedling 1.0.0. but not Ceedling 0.31

jerdwys commented:
@simeon-s1 I think you are missing the name change of the report.xml file that happened in Ceedling 1.0.0.  For me, it didn't report the results until I updated the xml report name.  Here is the change:
--- a/src/adapter.ts
+++ b/src/adapter.ts
@@ -647,13 +647,19 @@ export class CeedlingAdapter implements TestAdapter {

     private setXmlReportPath(projectKey: string, ymlProjectData: any = undefined) {
         let reportFilename = 'report.xml';
-        if (ymlProjectData) {
-            try {
-                const ymlProjectReportFilename = ymlProjectData[':xml_tests_report'][':artifact_filename'];
-                if (ymlProjectReportFilename != undefined) {
-                    reportFilename = ymlProjectReportFilename;
-                }
-            } catch (e) { }
+        if (this.isOldCeedlingVersion) {
+               if (ymlProjectData) {
+                   try {
+                       const ymlProjectReportFilename = ymlProjectData[':xml_tests_report'][':artifact_filename'];
+                       if (ymlProjectReportFilename != undefined) {
+                           reportFilename = ymlProjectReportFilename;
+                       }
+                   } catch (e) { }
+               }
+        }
+        else
+        {
+               reportFilename = 'cppunit_tests_report.xml';
         }
         this.reportFilenames[projectKey] = reportFilename;
     }

jerdwys committed:
Added ability to detect cppunit output file changes in project.xml

jerdwys committed:
Added mention of the report_tests_log_factory for Ceedling 1.0.0

gsmithufl commented:
> @gsmithufl I believe you need to enable cppunit under report_tests_log_factory.
> 
> From the [breaking changes](https://github.com/ThrowTheSwitch/Ceedling/blob/test/ceedling_0_32_rc/docs/BreakingChanges.md#consolidation-of-test-report-generation-plugins-%EF%B8%8F-report_tests_log_factory): In addition, all references and naming connected to the previous xml_tests_report plugin have been updated to refer to CppUnit rather than generic XML as this is the actual format of the report that is processed.

@jerdwys you're right, thanks! I needed to enable `cppunit` and also also needed to change the name of the report from the new `cppunit_tests_report.xml` name to the old `reports.xml` name. For anyone else the documentation on how to do that is here: https://github.com/ThrowTheSwitch/Ceedling/tree/test/ceedling_0_32_rc/plugins/report_tests_log_factory#configuration

```yml
:report_tests_log_factory:
  :reports:
    # - json
    # - junit
    - cppunit
    # - html
  :cppunit:
    :filename: 'report.xml'
```

I think the extension would need to be updated to capture the new report name.

simeon-s1 commented:
hey, could you do a PR? then i'll merge it and also merge everything into main afterwards.

simeon-s1 committed:
Merge pull request #1 from jerdwys/ceedling-1.0.0-report-filename-fix

Ceedling 1.0.0 report filename fix

mmerlo commented:
Thanks for everyone's effort. I've tested this extension and I am having a couple of issues with debugging.

The first issue relates to selecting the specific launch config for debugging.

My `launch.json` had the ceedling debugging profile set as `"name": "ceedlingExplorer"`. The workspace`settings.json` in `.vscode` folder has this entry: `"ceedlingExplorer.projects": [{"path": "./", "debugLaunchConfig": "ceedlingExplorer"}]`. When attempting to debug with these settings, vscode gives the message: `Configuration 'ceedling' is missing in 'launch.json'.` Changing the `launch.json` profile name to `ceedling` fixes the issue. Either I'm setting this profile name wrong in the settings or it is being ignored.

The second issue is that the test timer in Test Explorer continues to run if the debugging button is used to run tests. The stop button is the only way I have found to stop the timer. When in this condition (ie the timer is running after debug), the Explorer otherwise behaves as expected and tests can be run with the run buttons.

EmileBV commented:
Hi, I was just wondering if there are plans to complete this PR now that Ceedling 1.0.0 is out?

yahyayozo commented:
Any updates on this PR? it will be so helpful to support the Ceedling 1.0.0

Commit:
fix(adapter): improve error handling in project.yml file operations

- Enhance error messages to include the project.yml file path
- Add error logging for better debugging

Commit:
refactor(adapter): clear file watches and improve XML report deletion

- Add file watch clearing in dispose method to prevent memory leaks
- Update deleteXmlReport method to check for file existence before attempting deletion

simeon-s1 committed:
Merge pull request #2 from mdeweerd/msg/errors

chore: Improvements to error messages, dispose

Yangnaan committed:
Update adapter.ts

when debug finished, restore all buttons state to default.

jnz86 commented:
Six months later, just checking in! It would be great to get this merged in to use this extension again.

simeon-s1 committed:
Merge pull request #4 from Yangnaan/ceedling-1.0.0-compatibility-and-multi-project-setup

Update adapter.ts

lumal21 committed:
Update package.json, adapter.ts, and main.ts to improve functionality and fix issues

Create option to get full path of executable to debug.
Prefix ids with project key to avoid conflicts with file with same name in other projects.
Fix scope of projects variable to use config from settings.json on multi-root workspace.
Add option to ./ in shell path resolve to workspace root.

mmerlo commented:
Is this PR still needed? It looks like release v1.13 added Ceedling 1.0 compatibility.

simeon-s1 committed:
Merge pull request #5 from lumal21/fix-scope-of-projects-config

Update package.json, adapter.ts, and main.ts to improve functionality and fix issues
